codegen/debugger.go

```diff
@@ -0,0 +1,546 @@
+	d.sendControl(DebugNext)
+	return d.GetState()
+}
+
+func (d *debugger) Restart() (*State, error) {
+	d.sendControl(DebugRestart)
+	return d.GetState()
+}
+
+func (d *debugger) Step() (*State, error) {
+	d.sendControl(DebugStep)
+	return d.GetState()
+}
+
+func (d *debugger) StepOut() (*State, error) {
+	d.sendControl(DebugStepOut)
+	return d.GetState()
+}
+
+func (d *debugger) Backtrace() ([]Frame, error) {
+	s, err := d.GetState()
+	if err != nil {
+		return nil, err
+	}
+	return Backtrace(s.Ctx), nil
+}
+
+func (d *debugger) Breakpoints() []*Breakpoint {
+	return d.breakpoints
+}
+
+func (d *debugger) CreateBreakpoint(bp *Breakpoint) (*Breakpoint, error) {
+	if _, ok := d.breakpointIDs[bp.ID()]; ok {
+		return bp, fmt.Errorf("breakpoint already exists at %s", bp.ID())
+	}
+	bp.dbgr = d
+	d.breakpoints = append(d.breakpoints, bp)
+	d.breakpointIDs[bp.ID()] = struct{}{}
+	return bp, nil
+}
+
+func (d *debugger) ClearBreakpoint(bp *Breakpoint) error {
+	if _, ok := d.breakpointIDs[bp.ID()]; !ok {
+		return fmt.Errorf("breakpoint at %s does not exist", bp.ID())
+	}
+	if bp.SourceDefined {
+		return fmt.Errorf("cannot clear source defined breakpoint at %s", bp.ID())
+	}
+	for i, candidate := range d.breakpoints {
+		if candidate == bp {
+			d.breakpoints = append(d.breakpoints[:i], d.breakpoints[i+1:]...)
+			break
+		}
+	}
+	delete(d.breakpointIDs, bp.ID())
+	return nil
+}
+
+func (d *debugger) Terminate() {
+	// Set debugger error so that next yield it exits early.
+	d.err = ErrDebugExit
+	d.sendControl(DebugTerminate)
+}
+
+func (d *debugger) Exec(ctx context.Context, stdin io.ReadCloser, stdout, stderr io.Writer, args ...string) error {
+	s, err := d.GetState()
+	if err != nil {
+		return err
+	}
+
+	var se *errdefs.SolveError
+	if errors.As(s.Err, &se) {
+		var ge *gatewayError
+		if !errors.As(s.Err, &ge) {
+			return errors.Wrap(s.Err, "solve error without gateway")
+		}
+
+		// If the gateway context is canceled, also cancel the user's context.
+		ctx, cancel := context.WithCancel(ctx)
+		go func() {
+			<-ge.Done()
+			cancel()
+		}()
+
+		return ExecWithSolveErr(ctx, ge.Client, se, stdin, stdout, stderr, args...)
+	}
+
+	fs, err := s.Value.Filesystem()
+	if err != nil {
+		return err
+	}
+
+	return ExecWithFS(ctx, d.cln, fs, s.Options, stdin, stdout, stderr, args...)
+}
+
+func (d *debugger) sendControl(control DebugMode) {
+	// Prevent control being sent in parallel.
+	d.mu.Lock()
+	select {
+	case <-d.done:
+		// If debugger is closed, then release lock for other clients to exit
+		// gracefully.
+		d.mu.Unlock()
+		return
+	default:
+	}
+
+	// Otherwise send control, note that when multiple clients send control on
+	// the same state it will queue.
+	d.wg.Add(1)
+	defer d.wg.Done()
+	d.control <- control
+}
+
+func (d *debugger) Close() error {
+	// Set the debugger exit err.
+	d.err = ErrDebugExit
+	// Cancel incoming control signals.
+	close(d.done)
+	// Wait for clients to exit gracefully.
+	d.wg.Wait()
```

Comment:
It is already holding `d.mu` here, but the certainty is because of where `dbgr.Close` is called in `command/run.go`. Open to ideas on how to design this better so that developer error isn't a concern.